NE-1946: ingress: Add gateway-api-crd-life-cycle-management (#1740)
openshift/enhancements: enhancements/ingress/gateway-api-crd-life-cycle-management.md

---
title: gateway-api-crd-life-cycle-management
authors:
  - "@Miciah"
  - "@shaneutt"
reviewers:
  - "@shaneutt" # Gateway API maintainer.
  - "@dgn" # OSSM maintainer.
  - "@JoelSpeed" # SME for a similar issue and solution in CAPI.
  - "@knobunc" # Staff engineer for Networking, involved in past discussions.
approvers:
  - TBD
api-approvers:
  - None
creation-date: 2025-01-22
last-updated: 2025-01-22
tracking-link:
  - https://issues.redhat.com/browse/NE-1946
see-also:
  - "/enhancements/ingress/gateway-api-with-cluster-ingress-operator.md"
---

# Gateway API CRD Life-Cycle Management

This enhancement describes how the [Ingress Operator](https://github.com/openshift/cluster-ingress-operator) manages Gateway API
CRDs.

## Summary

A key goal of Gateway API is to nurture a diverse ecosystem of implementations.
These implementations, including OpenShift's, all rely on the same Gateway API
CRDs.  To these ends, OpenShift must manage the life-cycle of these CRDs in a
way that enables Red Hat to support the product, minimizes friction for the

> end-user, and avoids causing conflicts with other implementations that rely on

Reviewer: ```suggestion
end-user, provides consistency and avoids causing conflicts with other implementations that rely on
```

these CRDs.

## Motivation

Gateway API is an evolving API, and its CRDs change over time.  For core
OpenShift, we must ensure that we have a version of the CRDs that is compatible
with our product, and we must handle upgrading these CRDs for our own needs.  We
must also avoid unnecessarily causing friction for Red Hat layered products,
partner products, and other third-party implementations.

### User Stories

#### Using OpenShift's Gateway API implementation on a new OpenShift cluster

As a cluster-admin, I want to install a new OpenShift 4.19 cluster, and then use
OpenShift's Gateway API implementation to configure ingress to workload on this
cluster.

_TBD_

#### Using OpenShift's Gateway API implementation on an upgraded cluster

As a cluster-admin, I want to upgrade my cluster from OpenShift 4.18 (which
*doesn't* manage the Gateway API CRDs' life-cycle) to OpenShift 4.19 (which

> *does*).

Author: Yeah, I figured the user story could describe two cases: CRDs were not installed on the 4.18 cluster, and they were.  I can split it into two user stories if that makes more sense.

> *does*).

Reviewer 2: Yeah I think it might make sense to have one explicitly call out that there were Gateway API CRDs installed in 4.18

_TBD_

#### Future OpenShift upgrades

As a cluster-admin, I want to upgrade my cluster from OpenShift 4.19 (which
manages the Gateway API CRDs' life-cycle) to OpenShift 4.20 (which includes a
newer version of these CRDs).

_TBD_

### Goals

- Ensure Gateway API CRDs are installed on new and upgraded OpenShift 4.19 clusters.
- Ensure the installed Gateway API CRDs are compatible with OpenShift's needs.
- Detect and warn if an incompatible version of the CRDs is installed.
- Describe the issue of dead fields and potential risks thereof.

### Non-Goals

- Automatically replace incompatible CRDs that some other agent installed.
- Provide a pre-upgrade check in OpenShift 4.18 for incompatible CRDs.
- Provide an explicit override for the cluster-admin to take CRD ownership.
- Implement an admission webhook to block updates to dead fields.
- Solve CRD life-cycle management in OLM.
- Solve OLM subscription management.
- Solve CRD life-cycle management for OSSM or Istio resources.

## Proposal

### Life-cycle management

OpenShift's Ingress Operator monitors for the presence of the Gateway API CRDs.

- If they are absent, it installs the appropriate version.

> - If they are present at the expected version, the operator does nothing.

Reviewer: If "do nothing" also implies we will use these CRDs I think we'll need to think on that more because we need to manage their lifecycle, and we would need to know who the owner is and have that consent. This however I can add in one of the follow up iterations, so a TODO here should be good:

```suggestion
- If they are present at the expected version: TODO - expand upon what we'll do for preexisting CRDs.
```

Author: Are you alluding to transferring ownership with Server-Side Apply?  Now that you mention it, I think the proposal should be rewritten in terms of who owns it rather than whether it is at the expected version.  Should I go ahead and s/at the expected version/with the Ingress Operator as the owner/ now, with a to-do to fill in the SSA-related details?

Reviewer: Yes that sounds great :+1:

Author: I rewrote in terms of SSA and added a to-do in https://github.com/openshift/enhancements/compare/99aae59bb1ac50f1ef02e07ff3dcc3385216c451..e34441f0d0685989e6417db70dcf049c68f9fc55.

- If they are present at an unexpected version, the operator signals a degraded state.

Note that the appropriate version for the CRDs depends on the version of
OpenShift Service Mesh (OSSM) that the Ingress Operator has pinned in a given

> OpenShift release.  Specifically, the CRDs SHOULD be the version corresponding
> to the version of Istio in that OSSM version; they MUST be of some version that
> is compatible and that we have tested with that version of Istio.

Reviewer: ```suggestion
OpenShift release.  Specifically, the CRDs SHOULD be the version corresponding
to the version of Istio in that OSSM version; they MUST be of some version that
is compatible and that we have tested with that version of Istio (TODO: expand upon this)
```

Author: Done in https://github.com/openshift/enhancements/compare/99aae59bb1ac50f1ef02e07ff3dcc3385216c451..e34441f0d0685989e6417db70dcf049c68f9fc55.  Is that sufficient?  If we need more detail, it might warrant a separate section for how we validate a particular CRD version.

OpenShift, the cluster-admin might be advised to pay particular attention to
OSSM's resource requirements and the cluster's resource constraints before
attempting to use Gateway API.

MicroShift does not run the Ingress Operator and has its own design for
supporting Gateway API (see the MicroShift [gateway-api-support](../microshift/gateway-api-support.md)
enhancement).

### Implementation Details/Notes/Constraints

> What are some important details that didn't come across above in the
> **Proposal**? Go in to as much detail as necessary here. This might be
> a good place to talk about core concepts and how they relate. While it is useful
> to go into the details of the code changes required, it is not necessary to show
> how the code will be rewritten in the enhancement.

### Risks and Mitigations

#### Dead fields

If an installed Gateway API CRD includes some field that is not implemented by
the installed Gateway API implementation, then updates to this field may be
silently ignored.  In the context of a security feature, ignoring this field
could result in a configuration that inadvertently exposes workload.

Potential mitigation strategies:

> - Handle with an admission webhook.

Reviewer: (just a note that we put handling updates to dead fields being handled by a webhook as a non-goal above)

We have a dilemma when it comes to trying to handle dead fields with a webhook, because our validation would require us to ultimately reference other objects. Something we need to think through, but we can follow up on that as part of the TBD.

Author: I added a cross-reference to the "Admission webhook" section under "Alternatives", and I added a paragraph about consistency issues to that section.  I left the TBD to fill in more details.  These changes are in https://github.com/openshift/enhancements/compare/99aae59bb1ac50f1ef02e07ff3dcc3385216c451..e34441f0d0685989e6417db70dcf049c68f9fc55.

> - Work upstream to address this issue.

Reviewer: Upstream is in fact going to try and avoid unsafe dead field situations (e.g. security related fields) as much as is feasible, but in fact we know today that Auth fields are on their way soon and it wont be a major release cut to add them. Again something we can expand upon via the TBD, the mitigation with upstream is mostly going to be about awareness and preparation because we're tuned into upstream.

> - How will it be tested in isolation vs with other components?
> - What additional testing is necessary to support managed OpenShift service-based offerings?
> 
> No need to outline all of the test cases, just the general strategy. Anything
> that would count as tricky in the implementation and anything particularly
> challenging to test should be called out.
> 
> All code is expected to have adequate tests (eventually with coverage
> expectations).

The Ingress Operator will have E2E tests to simulate the user stories outlined
in this EP:

- Delete the CRDs; verify that the operator re-installs them.
- Install incompatible CRDs with `metadata.managedFields` set to indicate that the operator *did not* install them; verify that the operator reports the appropriate `Degraded` status.
- Install older CRDs with `metadata.managedFields` set to indicate that an older version of the operator *did* install them; verify that the operator updates them.

_TBD_

## Graduation Criteria

Note that Gateway API CRD life-cycle management will be part of the
[gateway-api-with-cluster-ingress-operator](gateway-api-with-cluster-ingress-operator.md) enhancement, which defines its

> own graduation criteria.

Reviewer: We had a conversation before about potentially packaging the CRDs in at the CVO level, we should capture somewhere in this doc (alternatives considered, etc) the reasons why we moved away from that for posterity.

Author: I added a heading in <https://github.com/openshift/enhancements/compare/99aae59bb1ac50f1ef02e07ff3dcc3385216c451..e34441f0d0685989e6417db70dcf049c68f9fc55>.  I don't actually understand how using CVO would help, so I haven't written anything out yet.  I can go back through Jira comments and meeting notes to try to figure that out, or else feel free to write something up about that alternative.

run the upstream conformance tests and downstream end-to-end tests with the
specific combination of OSSM version that the operator installs and CRD version
that is desired to be installed.

These CRDs are necessary for the Gateway API feature to work at all.  Ideally,
the Ingress Operator is the only agent that attempts to manage the life-cycle of
these CRDs (that is, creating, updating, or deleting them).  However, in some
contingencies, this is not the case:

- A cluster-admin can install the CRDs before upgrading to OpenShift 4.19.
- A layered product or third-party controller could be actively managing them.

In any scenario in which some agent other than the Ingress Operator has
installed the CRDs, the Ingress Operator cannot trivially infer the intent and
potential active use of the existing CRDs.  Overwriting the existing versions
represents a risk of breaking workload or violating the end-user's expectations.
Thus the Ingress Operator will only detect and warn about these scenarios.

The Ingress Operator's warning will take the form of, at a minimum, setting the
`Degraded` status condition's status to `True` with a descriptive message on the

> ingress clusteroperator.  

Reviewer 2: At some point we will need to explain how an admin gets out of this, TODO delete this comment if I find that later

Author: I figure it can be documented under "Workflow Description" or "Support Procedures".  I can add a cross-reference here.

Reviewer 2: That would be helpful, thank you

>   automatically in CI) and by whom (e.g. perf team; name the responsible person and let them review
>   this enhancement)
> 
> - Describe the possible failure modes of the API extensions.
> - Describe how a failure or behaviour of the extension will impact the overall cluster health
>   (e.g. which kube-controller-manager functionality will stop working), especially regarding
>   stability, availability, performance and security.
> - Describe which OCP teams are likely to be called upon in case of escalation with one of the failure modes
>   and add them as reviewers to this enhancement.

_TBD: Do we need to describe anything here?_

## Support Procedures

### Conflicting CRDs

If the Ingress Operator detects the presence of a conflicting version of the
Gateway API CRDs, it updates the ingress clusteroperator to report a `Degraded`
status condition with status `True` and a message explaining the situation:

_TBD: Insert example output from `oc get clusteroperators/ingress -o yaml`._

In this situation, the cluster-admin is expected to verify that workload would
not be broken by handing life-cycle management of the CRDs over to the Ingress

> Operator:

Reviewer 2: If they cannot verify this, or I guess, they verify it and determine their workload would be broken, we need to explain that path too

Author: Right, we will probably need to describe how workload could be broken (e.g. third-party controller fails to start because it needs newer CRDs), how to predict or detect it, and how to handle it (even if it comes down to, "Work with Red Hat and the third-party vendor to align on a common CRD version.").

_TBD: Insert `oc` command to make the CRD ownership transition._

Then the Ingress Operator takes ownership and updates the CRDs:

_TBD: Insert example `oc get clusteroperators` and `oc get crds` commands._

### Overriding the Ingress Operator

_TBD: Should we describe how to turn off the Ingress Operator so that the
cluster-admin can override the CRDs, or describe how Server-Side Apply enables

> the cluster-admin to take over the CRDs?_

Reviewer 2: Sounds like you might need what we are doing for CAPI in this space?

Author: I'm hoping we don't need to implement an explicit override.  Even if the alternative is something a little complicated (involving SSA) or heavy-handed (such as using a CVO override and scaling down the whole operator), we just need something good enough for now and defer doing something better.
